Add setting whether ACSL assertions should be checked

schuessf opened the pull request:
This PR adds a setting for `CACSL2BoogieTranslator` (in the group "specification") whether ACSL assertions should be checked or not. For compatibility, I enabled this setting by default, as previously ACSL was always checked.
Note: For now, we still parse ACSL even if the setting is disabled. This is an issue, if a comment contains unsupported ACSL (or starts with @ and contains no ACSL at all #667). But disabling parsing could be more complicated, as it is done in `CDTParser` instead of `CACSL2BoogieTranslator`.

schuessf committed:
Add setting whether ACSL assertions should be checked

maul-esel commented:
:+1: thanks for adding this.

Regarding the parsing issues, we've also had this issue that @bahnwaerter reported that other comments not meant to be ACSL (I think it was doxygen) lead to parse errors. How should we handle this?
- Should we add a separate setting to CDTParser to (in the class `CommentParser`) skip parsing of ACSL?
  + Downside: This other setting would effectively override the setting added in this PR. That could lead to confusing behaviour.
- In case we cannot parse a particular comment (that we believe to be ACSL), instead of directly reporting a syntax error, could we instead annotate the C AST with a kind of "problem marker" (instead of an ACSL AST), and only report a syntax error if CACSL2BoogieTranslator encounters the problem marker?

schuessf commented:
> In case we cannot parse a particular comment (that we believe to be ACSL), instead of directly reporting a syntax error, could we instead annotate the C AST with a kind of "problem marker" (instead of an ACSL AST), and only report a syntax error if CACSL2BoogieTranslator encounters the problem marker?

I think this is a good idea, I will check how to implement this approach.

schuessf committed:
Add ACSLProblemNode to crash only if ACSL should be checked

Before, we always canceled the toolchain, if an error during parsing of ACSL occured (see CommentParser). However,
since ACSL checks can now be disabled, we should not crash if they are disabled. Instead, we return an ACSLProblemNode
if parsing fails and we only throw an exception when handling such anACSLProblemNode in ACSLHandler, which is only
called if ACSL should be checked.

schuessf commented:
Okay, I implemented it this way: Instead of throwing reporting an error and cancelling the toolchain in `CommentParser`, we just create an `ACSLProblemNode`. Now we only crash when dispatching such an `ACSLProblemNode` in `ACSLHandler`, i.e., only if the new setting `Check ACSL annotations` is enabled.

Also, I just again, if this setting affects witness validation, where we also make use of our ACSL parsing. Witness validation still works as expected, even if `Check ACSL annotations` is disabled, since the dispatcher is called directly. And if the witness contains a syntax error, we still crash as before.

maul-esel commented on trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/ACSLHandler.java:
```diff
@@ -950,4 +951,9 @@ public Result visit(final IDispatcher main, final NullPointer node) {
+	@Override
+	public Result visit(final IDispatcher main, final ACSLProblemNode node) {
+		final ILocation loc = mLocationFactory.createACSLLocation(node);
+		throw new UnsupportedSyntaxException(loc, "Error during parsing of ACSL (" + node.getErrorMessage() + ")");
```
How is this syntax exception handled at runtime? How does the behaviour compare to what happened before these changes (do/did we get generic `ExceptionOrErrorResult`s, or some kind of `UnsupportedSyntaxResult`, etc.) ? And how does it relate to the `catch (final ParseException e1)` in line 2583 of `CHandler`?

danieldietsch replied:
This will be catched and turned into an UnsupportedSyntaxResult.

schuessf replied:
> How is this syntax exception handled at runtime? How does the behaviour compare to what happened before these changes (do/did we get generic ExceptionOrErrorResults, or some kind of UnsupportedSyntaxResult, etc.) ?

The result / output does not change much. The only difference is that we get an `UnsupportedSyntaxResult` (the exception is catched in [`MainTranslator`](https://github.com/ultimate-pa/ultimate/blob/dev/trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/MainTranslator.java#L124)) instead of a `SyntaxErrorResult`, but this is this result that is usually produced in the C translation.
Before:
```
WARN  L194          CommentParser]: Syntax Error: #45[gt](0/12 - 0/13)
[...]
INFO  L172             ResultUtil]:  --- Results ---
[...]
INFO  L174             ResultUtil]:  * Results from CDTParser:
INFO  L198             ResultUtil]:   - SyntaxErrorResult [Line: 3]: Incorrect Syntax
INFO  L202             ResultUtil]:     Syntax Error: #45[gt](0/12 - 0/13)
```
After:
```
ERROR L321         MainTranslator]: Unsupported Syntax: Error during parsing of ACSL (Syntax Error: #45[gt](0/12 - 0/13))
[...]
INFO  L172             ResultUtil]:  --- Results ---
[...]
INFO  L174             ResultUtil]:  * Results from de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator:
INFO  L198             ResultUtil]:   - UnsupportedSyntaxResult [Line: 3]: Unsupported Syntax
INFO  L202             ResultUtil]:     Error during parsing of ACSL (Syntax Error: #45[gt](0/12 - 0/13))
```

schuessf replied:
> And how does it relate to the catch (final ParseException e1) in line 2583 of CHandler?

`IDispatcher::nextACSLStatement` is allowed to throw a `ParseException`, but the actual implementations never do so. I guess `ParseException` was used earlier and this is just a leftover. But I am not sure, if the exception handling for `ParseException` should be removed in this PR.

schuessf replied:
I just removed `ParseException` in 9116035 (on dev).

schuessf commented:
I just rebased this PR again, can we merge it, or do you want some changes (@maul-esel @Heizmann @bahnwaerter @danieldietsch)?

danieldietsch commented:
Merge :)
